test/gtest/uct/ib/test_devx.cc

```diff
@@ -48,4 +48,202 @@ UCS_TEST_P(test_devx, dbrec)
+        void *addr                     = ucs_malloc(length, "test");
+        uct_md_mem_reg_params_t params = {};
+        uct_mem_h memh;
+
+        ASSERT_UCS_OK(uct_ib_mlx5_devx_mem_reg(m_e->md(), addr, length, &params,
+                                               &memh));
+        return ucs_derived_of(memh, uct_ib_mlx5_devx_mem_t);
+    }
+
+    void destroy_memh(uct_ib_mlx5_devx_mem_t *memh) const {
+        void *addr                       = memh->address;
+        uct_md_mem_dereg_params_t params = {};
+
+        params.field_mask = UCT_MD_MEM_DEREG_FIELD_MEMH;
+        params.memh       = (uct_mem_h)memh;
+
+        ASSERT_UCS_OK(uct_ib_mlx5_devx_mem_dereg(m_e->md(), &params));
+
+        if (nullptr != addr) {
+            ucs_free(addr);
+        }
+    }
+
+    void check_umr_init(bool init_export, bool init_import) const {
+        EXPECT_EQ(init_export, nullptr != md()->umr.qp);
+        EXPECT_EQ(init_export, nullptr != md()->umr.cq);
+        if (!init_export) {
+            EXPECT_TRUE(ucs_list_is_empty(&md()->umr.mkey_pool));
+        }
+
+        EXPECT_EQ(init_import, nullptr != md()->umr.mkey_hash);
+    }
+
+    static void check_memh_export(uct_ib_mlx5_devx_mem_t *memh, bool is_umr) {
+        EXPECT_EQ(is_umr, UCT_IB_MLX5_MKEY_TAG_UMR ==
+                          (memh->exported_lkey & 0xff));
+        EXPECT_EQ(is_umr, nullptr != memh->exported_umr_mkey);
+        EXPECT_EQ(is_umr, nullptr == memh->cross_mr);
+    }
+};
+
+UCS_TEST_P(test_devx_umr_mkey, lazy_init)
+{
+    check_umr_init(false, false);
+}
+
+UCS_TEST_P(test_devx_umr_mkey, export_mkey_no_xgvmi)
+{
+    uct_ib_mlx5_devx_mem_t *memh = create_memh(1024);
+
+    /* Drop XGVMI capability */
+    md()->flags &= ~UCT_IB_MLX5_MD_FLAG_INDIRECT_XGVMI;
+
+    ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), memh));
+
+    /* UMR objects remains uninitialized, exported key is not UMR one */
+    check_umr_init(false, false);
+    check_memh_export(memh, false);
+    destroy_memh(memh);
+}
+
+UCS_TEST_P(test_devx_umr_mkey, export_mkey_xgvmi)
+{
+    if (!(md()->flags & UCT_IB_MLX5_MD_FLAG_INDIRECT_XGVMI)) {
+        UCS_TEST_SKIP_R("XGVMI is not supported");
+    }
+
+    uct_ib_mlx5_devx_mem_t *memh = create_memh(1024);
+    ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), memh));
+
+    /* Check that XGVMI capability had not been dropped */
+    ASSERT_TRUE(md()->flags & UCT_IB_MLX5_MD_FLAG_INDIRECT_XGVMI);
+
+    /* UMR export transport objects must be initialized */
+    check_umr_init(true, false);
+    /* UMR mkey pool must be empty */
+    ASSERT_TRUE(ucs_list_is_empty(&md()->umr.mkey_pool));
+    check_memh_export(memh, true);
+    destroy_memh(memh);
+
+    /* After release of memh, associated UMR mkey is moved back to mkey pool */
+    ASSERT_EQ(1, ucs_list_length(&md()->umr.mkey_pool));
+}
+
+UCS_TEST_P(test_devx_umr_mkey, export_mkey_xgvmi_reuse)
+{
+    if (!(md()->flags & UCT_IB_MLX5_MD_FLAG_INDIRECT_XGVMI)) {
```

Comment:
Yes, indeed that is a good option!
I tried that initially, but if failed with exception, because `m_e` was created only in `init`, but this check-for-skip-test is called after object constructor but before `init`. So I moved `m_e` object creation to the ctor and it works like that

Reply:
Yes, this is exactly what I've implemented

Reply:
i see, i misunderstood you last comment.